src/http/client.cc

```diff
@@ -164,13 +163,14 @@ future<connection::reply_ptr> connection::do_make_request(request req) {
                 });
             });
         });
-    });
 }
 
 future<reply> connection::make_request(request req) {
-    return do_make_request(std::move(req)).then([] (reply_ptr rep) {
+  return do_with(std::move(req), [this] (auto& req) {
```

Comment:
could you please extract this into its own commit? as it actually prolongs the lifecycle of `req` so that it is alive until `do_make_request(req)` is resolved ? this addresses the problem we discussed the other day on why `s3::client::get_object_tagging()` does not crash. and with this change, i think we will be able to drop the tricky hacks like https://github.com/scylladb/scylladb/blob/2a81b2e49af8ca01a058675dd4b468c6145f9aad/utils/s3/client.cc#L271 . as `req` as the captured variables will be alive after the lambda resumes after it is suspended, even we don't capture it as a local variable and force the coroutine runtime to save it to the coroutine frame.

Reply:
so my impression was that there two things going on here:

1. change the parameter type
2. use `do_with()`

Reply:
i see. could you please emphasis the problem addressed by this change, and explain how the problem is addressed?

> Now the method in question gets the request value and runs its code inside the do_with(std::move(request), ...) context

is not obvious to me that "now ... gets ..." is the change made in this very commit. it is just not explicit enough, i think.

and the problem addressed by keeping the `request` alive using `do_with` , and the change of `do_with()` is missing.

would you mind addressing these concerns?

Reply:
Well, the "problem" is: there are two methods:

- `connection::make_request()`
- `connection::do_make_request()`

The former _moves_ request into the latter. Also, there's `client::make_request()` which also _moves_ request into `connection::do_make_request()`. 

Next, this PR tries to teach `client::make_request()` restart requests. Restarting means that `client::make_request()` calls `connection::do_make_request()`. If the latter fails, the former needs to do something and (!) call it _again_. But when calling it again it should have the request at hand to pass it as an argument. In current approach `client::make_request()` won't have the request, because it had _moved_ it into the `connection::do_make_request()` on the first try. So I need to keep request object in `client::make_request()` so that it does _not_ move it into `connection::do_make_request()` but passes there a reference. Reference can be passed several times (and #1848 makes sure request remains intact between retries)

Reply:
thank you for the detailed explanation. even the change is a necessity of the retry implementation, it still addresses the problem we discussed couple weeks ago. personally, i would want to ensure that i can drop the hackish local variables in the caller, after reading the commit message. but i cannot find it in the commit message. so i am wondering if it is possible to note it in the commit message.

and i think, my concern is the impaction of the change, which _happens_ to address the problem we discussed the other day. not quite related to the "problem" you explained in your previous reply.